### ISSUE TITLE
on constraints.distribute(solution) after solution_trans.interpolate within the function refine_mesh

### ISSUE THREAD

**Issue by jixiang7** [dealii/dealii#757]
In step-26, we did not call constraints.distribute(solution) after calling solution_trans.interpolate. Should we add it?

**Comment by bangerth** [comment]
Yes, we should. @jixiang7 : Want to give this a try yourself? ;-)

**Comment by jixiang7** [comment]
I compared the final results after adding constraints.distribute(solution). There is only a very small difference in results.

I did a test and found the size of solution vector does not change after calling distribute but ordering changes. Then I checked the online documentation and it only says "Re-distribute the elements of the vector in-place." So could you explain to me what this function is doing and why we need to change the ordering?

**Comment by jixiang7** [comment]
@bangerth

**Comment by tjhei** [comment]
```distribute(solution)``` sets constrained entries to the correct values. The size of the vector or the ordering is not going to change. The call to ```distribute``` needs to happen after ```interpolate``` because otherwise the solution might not satisfy boundary conditions or hanging node constraints exactly. For example boundary values of refined cells will be set to the interpolated boundary value and not the correct one coming from the boundary conditions.

**Comment by jixiang7** [comment]
@tjhei Thanks for clarifying it.

**Comment by bangerth** [comment]
See also #785

**Comment by jixiang7** [comment]
@tjhei As you said, the distribute(solution) is to satisfy the boundary conditions or constraints. Then I still do not understand why we also call hanging_node_constraints.distribute (newton_update) in step-15? The newton_update is not supposed to satisfy boundary conditions but zero.

**Comment by tjhei** [comment]
This is done so that the newton update satisfies the hanging nodes constraints exactly.

**Comment by bangerth** [comment]
#787 adds the missing call in step-26. But I think that a similar call is also missing in steps 31, 32, 33, 42, 43. It is there in step-15.

**Comment by bangerth** [comment]
@jixiang7 : The call adjusts the solution so that it satisfies whatever constraints are represented in the `ConstraintMatrix` object. In both step-15 and step-26, these are hanging node constraints but not boundary values. One can, however, also represent boundary values with `ConstraintMatrix` objects -- but step-15 and 26 just choose to not do this.
